[core] `AssetsDefinition.map_asset_specs` fails when attempting to add deps with `ins` set

### What's the issue?

The error occurs because `InputDefinition` objects don't have a `to_definition()` method,
but `OpDefinition.__init__` tries to call it when recreating the op with new inputs.

Output: 

```
>       my_asset.map_asset_specs(lambda s: new_spec)

dagster_bug_repro.py:32:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
.venv/lib/python3.10/site-packages/dagster/_core/definitions/assets/definition/assets_definition.py:1411: in map_asset_specs
    return replace_specs_on_asset(self, mapped_specs)
.venv/lib/python3.10/site-packages/dagster/_core/definitions/assets/definition/assets_definition.py:2059: in replace_specs_on_asset
    "node_def": assets_def.op.with_replaced_properties(
.venv/lib/python3.10/site-packages/dagster/_core/definitions/op_definition.py:387: in with_replaced_properties
    return OpDefinition.dagster_internal_init(
.venv/lib/python3.10/site-packages/dagster/_core/definitions/op_definition.py:211: in dagster_internal_init
    return OpDefinition(
.venv/lib/python3.10/site-packages/dagster/_core/decorator_utils.py:198: in wrapped_with_pre_call_fn
    return fn(*args, **kwargs)
.venv/lib/python3.10/site-packages/dagster/_core/definitions/op_definition.py:141: in __init__
    input_defs = [
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

.0 = <list_iterator object at 0x7f990378b010>

    input_defs = [
>       inp.to_definition(name) for name, inp in sorted(ins.items(), key=lambda inp: inp[0])
    ]  # sort so that input definition order is deterministic
E   AttributeError: 'InputDefinition' object has no attribute 'to_definition'

.venv/lib/python3.10/site-packages/dagster/_core/definitions/op_definition.py:142: AttributeError
```

### What did you expect to happen?

It should work correctly

### How to reproduce?

```py
import dagster as dg
import pytest


class TestMapAssetSpecsWithIns:
    """Bug: map_asset_specs fails when adding deps to assets with ins= parameter."""

    def test_adding_dep_to_asset_with_ins_fails(self):
        """Adding a new dep to an asset created with ins= raises AttributeError."""

        @dg.asset(ins={"my_input": dg.AssetIn(key="upstream")})
        def my_asset(my_input):
            pass

        spec = list(my_asset.specs)[0]
        new_spec = spec.replace_attributes(
            deps={*spec.deps, dg.AssetDep("another_upstream")},
        )

        # This should work but fails with:
        # AttributeError: 'InputDefinition' object has no attribute 'to_definition'
        with pytest.raises(AttributeError, match="to_definition"):
            my_asset.map_asset_specs(lambda s: new_spec)

    def test_adding_dep_to_asset_with_deps_only_works(self):
        """Adding a new dep to an asset with deps= (no function params) works fine."""

        @dg.asset(deps=["upstream"])
        def my_asset():  # No function parameters
            pass

        spec = list(my_asset.specs)[0]
        new_spec = spec.replace_attributes(
            deps={*spec.deps, dg.AssetDep("another_upstream")},
        )

        # This works because the function has no parameters
        result = my_asset.map_asset_specs(lambda s: new_spec)
        result_spec = list(result.specs)[0]
        dep_keys = {dep.asset_key for dep in result_spec.deps}

        assert dg.AssetKey("upstream") in dep_keys
        assert dg.AssetKey("another_upstream") in dep_keys
```

### Dagster version

1.12.3

### Deployment type

Local

### Deployment details

_No response_

### Additional information

_No response_

### Message from the maintainers

Impacted by this issue? Give it a 👍! We factor engagement into prioritization.

--- a/python_modules/dagster/dagster/_core/definitions/op_definition.py
+++ b/python_modules/dagster/dagster/_core/definitions/op_definition.py
@@ -142,7 +142,7 @@ class OpDefinition(NodeDefinition, IHasInternalInit):
         input_defs = [
             inp if isinstance(inp, InputDefinition) else inp.to_definition(name)
             for name, inp in sorted(ins.items(), key=lambda inp: inp[0])
-        ]   # sort so that input definition order is deterministic
+        ]  # sort so that input definition order is deterministic
 
         if isinstance(compute_fn, DecoratedOpFunction):
             resolved_input_defs: Sequence[InputDefinition] = resolve_checked_op_fn_inputs(

--- a/python_modules/dagster/dagster_tests/definitions_tests/test_asset_spec.py
+++ b/python_modules/dagster/dagster_tests/definitions_tests/test_asset_spec.py
@@ -253,27 +253,25 @@ def test_map_asset_specs_additional_deps() -> None:
 
 
 def test_map_asset_specs_asset_with_ins_regression() -> None:
-    """
-    This regression test prevents the recurrence of the AttributeError from OpDefinition.__init__.
-    """
-    
+    """This regression test prevents the recurrence of the AttributeError from OpDefinition.__init__."""
+
     @dg.asset(ins={"my_input": dg.AssetIn(key="upstream")})
     def my_asset(my_input):
         pass
 
     assets = [my_asset]
     spec = next(iter(my_asset.specs))
-    
+
     new_spec = spec.replace_attributes(
         deps={*spec.deps, dg.AssetDep("another_upstream")},
     )
 
     mapped_assets = dg.map_asset_specs(lambda s: new_spec, assets)
-    
+
     mapped_asset = mapped_assets[0]
     mapped_spec = next(iter(mapped_asset.specs))
     dep_keys = {dep.asset_key for dep in mapped_spec.deps}
-    
+
     assert dg.AssetKey("upstream") in dep_keys
     assert dg.AssetKey("another_upstream") in dep_keys
     assert mapped_asset.keys_by_input_name["my_input"] == dg.AssetKey("upstream")

